### DISCUSSION TITLE
Code changes in dependencies

### DISCUSSION THREAD

**Discussion by pjljvandelaar** [alire-project/alire#1198] [Q&A] [answered]
Dear Alire developers,

Sometimes when debugging a bug of mine, it is helpful to add some `put_line` statements in gnat libraries to better understand what is going on. When such a statement is added, I expect alire to rebuild at least the changed file.

For example, given the [bug_native12](https://github.com/AdaCore/libadalang/files/9476628/bug_native12.zip) project,
changing `C:\bug_native12\alire\cache\dependencies\gnatcoll_22.0.0_620c2f23\src\gnatcoll-projects.adb` by putting
` Put_Line ("Before Non_Blocking_Spawn");` before the call to `Non_Blocking_Spawn` on line 8266,
causes `alr run` to execute as follows
```
Note: Building bug_native12/bug_native12.gpr...
gnatcoll.gpr:254:07: warning: value "windows" is not used as label
Compile
   [Ada]          gnatcoll-scripts-projects.adb
   [Ada]          gnatcoll-projects.adb
   [Ada]          gnatcoll-projects-normalize.adb
   [Ada]          gnatcoll-projects-aux.adb
Build Libraries
   [gprlib]       gnatcoll.lexch
   [archive]      libgnatcoll.a
   [index]        libgnatcoll.a
   [gprlib]       gnatcoll_gmp.lexch
   [gprlib]       gnatcoll_iconv.lexch
   [archive]      libgnatcoll_gmp.a
   [archive]      libgnatcoll_iconv.a
   [index]        libgnatcoll_gmp.a
   [index]        libgnatcoll_iconv.a
   [gprlib]       langkit_support.lexch
   [bind SAL]     langkit_support
   [Ada]          b__langkit_support.adb
   [objcopy]      p__langkit_support_0.o
   [archive]      liblangkit_support.a
   [index]        liblangkit_support.a
   [gprlib]       adalang.lexch
   [bind SAL]     adalang
   [Ada]          b__adalang.adb
   [objcopy]      p__adalang_0.o
   [archive]      libadalang.a
   [index]        libadalang.a
Bind
   [gprbind]      bug_native12.bexch
   [Ada]          bug_native12.ali
Link
   [link]         bug_native12.adb
Build finished successfully in 82.72 seconds.
Start
Before Non_Blocking_Spawn
PS C:\bug_native12>
```
This is exactly as I expected, since I changed `gnatcoll-projects.adb` and some file depend on that file.

However, when I follow the call from `gnatcoll-projects.adb`  of `Non_Blocking_Spawn` in GNATStudio,
I end up in the file `C:\Users\laarpjljvd\.config\alire\cache\dependencies\gnat_native_11.2.4_2f9c5d6d\lib\gcc\x86_64-w64-mingw32\11.2.0\adainclude\g-expect.adb`. When I add a `put_line` statement in this file, I get
```
PS C:\bug_native12> alr run
Note: Building bug_native12/bug_native12.gpr...
gnatcoll.gpr:254:07: warning: value "windows" is not used as label
gprbuild: "bug_native12.exe" up to date
Build finished successfully in 3.44 seconds.
Start
Before Non_Blocking_Spawn
PS C:\bug_native12>
```

This is not what I expected! 
Are my expectations wrong? If so, please explain!

Thanks in advance for your answers!

Pierre

P.S.      Note the dependencies are resolved on my machine as follows:
```
   +i gnat            11.2.4 (new,installed,gnat_native,indirect)
   +  gnatcoll        22.0.0 (new,indirect)
   +  gnatcoll_gmp    22.0.0 (new,indirect)
   +  gnatcoll_iconv  22.0.0 (new,indirect)
   +  langkit_support 22.0.0 (new,indirect)
   +  libadalang      22.0.0 (new)
   +  libgmp          6.2.1  (new,indirect)
   +  libgpr          22.0.0 (new,indirect)
   +  libiconv        1.17.0 (new,indirect)
   +  xmlada          22.0.0 (new,indirect)
```
so I am a little bit surprised that I end up in gnat native 11.2.4.
Not about the number (although the project name hints at `native 12`)
but since gnat native is not in the dependencies list!

**Comment by Fabien-Chouteau** [comment, marked as answer]
`C:\Users\laarpjljvd\.config\alire\cache\dependencies\gnat_native_11.2.4_2f9c5d6d\lib\gcc\x86_64-w64-mingw32\11.2.0\adainclude\g-expect.adb` is a run-time file. Alire/gprbuild cannot recompile run-time files.
